### PR TITLE
feat: support ESLint v8

### PR DESCRIPTION
**Pull request by ota-meshi**
This PR updates this plugin to support ESLint v8.

However, currently eslint is still v8.0.0-beta-0.

**Summary of changes**

- ~~Change to get ESLint core rules from Linter instance.~~ See https://github.com/typescript-eslint/typescript-eslint/pull/3737#issuecomment-899159390.
- Adde `meta.hasSuggestions` to the rules that provide suggestions.
- Fix `no-dupe-class-members` and `no-extra-semi` to work with v8.0.0-beta.0.
- Change `CLIEngine` to be `undefined` in eslint v8.
- Add `jest-resolver.js`
  jest doesn't seem to support the exports map in package.json at this time.
  For this reason, jest-resolver.js is added as a workaround.
  See https://github.com/facebook/jest/issues/9771.

**About Lint errors:**
Since eslint-plugin-import and eslint-plugin-jest do not yet support eslint v8, still get a run error.

**Test failed on Node 10.x:**
Test fails because eslint v8 doesn't support Node 10.

Should I change the problematic workflow and use eslint v7 only for Node 10.x testing and linting? Or should I delete the workflow for Node 10.x?

### CODE REVIEW

**Comment by coderaiser** [comment]
That's amazing! This is the thing I wrote in #3735. The only thing that is missing is [use-at-own-risk](https://github.com/typescript-eslint/typescript-eslint/pull/3736/files) from [eslint migration](https://eslint.org/docs/8.0.0/user-guide/migrating-to-8.0.0#the-lib-entrypoint-has-been-removed). You can cherry-pick from my PR. And I think `node v10` can be removed from `CI` because it is not supported by `ESLint v8`.

File: packages/eslint-plugin/src/util/getEslintCoreRule.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  'no-unused-expressions': typeof import('eslint/lib/rules/no-unused-expressions');
+  'no-useless-constructor': typeof import('eslint/lib/rules/no-useless-constructor');
+  'no-restricted-globals': typeof import('eslint/lib/rules/no-restricted-globals');
+  'object-curly-spacing': typeof import('eslint/lib/rules/object-curly-spacing');
+  'prefer-const': typeof import('eslint/lib/rules/prefer-const');
+  quotes: typeof import('eslint/lib/rules/quotes');
+  semi: typeof import('eslint/lib/rules/semi');
+  'space-infix-ops': typeof import('eslint/lib/rules/space-infix-ops');
+  strict: typeof import('eslint/lib/rules/strict');
+}
```
Comment (coderaiser) [inline packages/eslint-plugin/src/util/getEslintCoreRule.ts:39]: This should not work, `eslint` [forbids](https://eslint.org/docs/8.0.0/user-guide/migrating-to-8.0.0#the-lib-entrypoint-has-been-removed) import from `lib`

**Comment by ota-meshi** [comment]
I don't think use-at-own-risk can be used unless this plugin drops support for eslint <= v7.

File: packages/eslint-plugin/src/util/getEslintCoreRule.ts (hunk 1 above)
Comment (ota-meshi) [reply packages/eslint-plugin/src/util/getEslintCoreRule.ts:39]: This is type-only import, so it works fine.

**Comment by bradzacher** [comment]
> I don't think use-at-own-risk can be used unless this plugin drops support for eslint <= v7.

It's possible for us to use both - it just means we can't migrate to ESM any time soon (not that we were planning on doing that. There's only one module we depend on that's made that switch so far).

This example code should work and allow us to support both, I believe:
```ts
import { version } from 'eslint';
import semver from 'semver';

const isESLintV8 = semver.major(version) >= 8;

export function getESLintCoreRule<R extends RuleId>(ruleId: R): RuleMap[R] {
  if (isESLintV8) {
    return require('eslint/use-at-own-risk').builtinRules[ruleId];
  } else {
    return require(`eslint/lib/rules/${ruleId}`);
  }
}

// for no-loss-of-precision
export function maybeGetESLintCoreRule<R extends RuleId>(ruleId: R): RuleMap[R] | null {
  try {
    return getESLintCoreRule<R>(ruleId);
  } catch {
    return null;
  }
}
```

**Comment by ota-meshi** [comment]
Thank you for the feedback! I changed to the method you suggested.

https://github.com/typescript-eslint/typescript-eslint/pull/3737/commits/21c76a332eb4be4e7caf647680ff6bbb3d3e9757#diff-65ca27a27fffac22b6d171704e8125ad44f370ae4d5b1f92f8d99ba612fa12b1R41-R58

**Comment by ota-meshi** [comment]
Do you have any ideas on how to fix a lint crash?

What do you think about changing to using eslint v7 in devDependencies and using eslint v8 only for CI unit tests?
This plugin is so popular that I think it could deadlock if you wait for other plugins to be fixed.

**Review by bradzacher — APPROVED** [review]
I pushed a fix for the broken linting - this was due to breaking changes not being supported yet. For now I've manually patched the packages to handle this. Eventually they'll release full support and we can drop those patches.

---

LGTM - thanks for this!

**Pull request merged**